### PR TITLE
Fix IBM i Db2 DSN parameters for local

### PR DESCRIPTION
**Pull request by jbh**
|      Q       |   A
|------------- | -----------
| Type         | bug
| Fixed issues |n/a

#### Summary

This fixes a bug, for lack of a better term, that arises when trying to use Doctrine locally on an IBM i system. There is a workaround, but that workaround isn't consistent with other database configurations. The IBMDB2/Driver.php uses IBMDB2/DataSourceName to translate any config into a connection string to pass to the db2_connect() and db2_pconnect() functions. When using this behavior remotely, the Db2 connect functions call the underlying non-IBM-i-system driver which takes the typical DSN parameters of DATABASE, HOSTNAME, PORT, PROTOCOL, UID, and PWD. However, on an IBM i, the Db2 connect functions call IBM i's very own [SQLDriverConnect](https://tinyurl.com/sdc-usage), which only takes DSN, UID, and PWD as arguments.

In summary, this checks if the IBMDB2/DataSourceName.php is running on an IBM i or not. If it is, it will build a connection string including DSN, UID, and PWD. Otherwise, it builds a typical connection string with DATABASE, HOSTNAME, PORT, PROTOCOL, UID, and PWD.

### CODE REVIEW

**Comment by jbh** [comment]
For anyone perhaps running into this issue while this PR is being reviewed or have in the past and gave up, below is the current workaround. I just didn't think it made sense to put it in the commit.

Seeing as Doctrine allows you to define your own DSN connection string in the dbname parameter (see [line #55](https://github.com/doctrine/dbal/blob/4.0.x/src/Driver/IBMDB2/DataSourceName.php#L55) of IBMDB2/DataSourceName.php), you can simply make it proper in your config. YAML and URL examples below.

```yaml
doctrine:
    dbal:
        driver: ibm_db2
        dbname: DSN=dbname;UID=username;PWD=password;
        host: localhost
```

```
ibm-db2://localhost/DSN=dbname;UID=username;PWD=password;?serverVersion=7.5.0
```

This works, but as stated in the commit, it's a requirement here where it's not in the other diver configurations.

**Comment by derrabus** [comment]
We don't support DB2 on IBM i. Without a CI, we simply cannot maintain it. 😞

**Comment by jbh** [comment]
Well, this patch should affect no one else, and it makes DBAL and ORM work on IBM i. I understand the need for CI on LUW, but again, this patch should have 0 affect on them and they'll keep passing fine.

**Comment by derrabus** [comment]
This is by far not the only difference between DB2 on x86 and IBM i. If you want to build a driver for System i, please do so in an external repository. We cannot have an opinion on the correctness of any patches for System i.

See also https://github.com/SeidenGroup/doctrine-dbal-ibmi

**Comment by jbh** [comment]
Well, works fine for me without SG's ODBC driver, which is mainly different due to using ODBC, not necessarily because it's IBM i. I'll just keep using the workaround. At least that's documented here.

Thanks.

**Comment by derrabus** [comment]
Thank you for understanding. 🙂
